### PR TITLE
[Feature] Support C2D

### PR DESCRIPTION
**Pull request by InvincibleWyq**
## Modification

Support C2D.
ps: ckpt & log & acc will be provided on Tuesday.

### CODE REVIEW

**Comment by ly015** [comment]
@InvincibleWyq Thank you very much for contributing to MMAction2!
There seems to be a lint issue. Could you please help fix it by running pre-commit hooks locally? You can configure the pre-commit hooks following https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md#code-style

File: mmaction/models/backbones/c2d_r50.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import torch
+import torch.nn as nn
+from mmcv.cnn import ConvModule
+
+from mmaction.models.backbones.resnet import ResNet
+from mmaction.registry import MODELS
+
+
+@MODELS.register_module()
+class C2D_R50(ResNet):
```
Comment (Dai-Wenxun) [inline mmaction/models/backbones/c2d_r50.py:13]: It is not necessary to limit the structure to `R50`,  a `R101` counterpart can be built in the same way, so it's better to change `C2D_R50` to `C2D`.

File: configs/_base_/models/c2d_r50.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        dropout_ratio=0.5,
+        init_std=0.01,
+        average_clips='prob'),
+    data_preprocessor=dict(
+        type='ActionDataPreprocessor',
+        mean=[123.675, 116.28, 103.53],
+        std=[58.395, 57.12, 57.375],
+        format_shape='NCTHW'),
+    train_cfg=None,
+    test_cfg=None)
```
Comment (Dai-Wenxun) [inline configs/_base_/models/c2d_r50.py:21]: It's better to delete these two lines.

File: configs/_base_/models/c2d_r50.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+model = dict(
+    type='Recognizer3D',
+    backbone=dict(
```
Comment (Dai-Wenxun) [inline configs/_base_/models/c2d_r50.py:3]: add `depth`.

File: configs/recognition/c2d/README.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+Specifically:
+
+- maxpool3d_1 = nn.MaxPool3d(kernel_size=(1, 3, 3), stride=(1, 2, 2), padding=(0, 0, 0))
+- maxpool3d_2 = nn.MaxPool3d(kernel_size=(2, 1, 1), stride=(2, 1, 1), padding=(0, 0, 0))
+
+C2D_Nopool implementation in MMAction2 is kept same as the ["SlowFast" repo](https://github.com/facebookresearch/SlowFast/blob/main/configs/Kinetics/c2/C2D_NOPOOL_8x8_R50.yaml)
+
+<!-- [ABSTRACT] -->
+
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:19]: add ABSTRACT of [Non-local Neural Networks](https://arxiv.org/abs/1711.07971)

File: configs/recognition/c2d/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+<!-- [IMAGE] -->
+
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:21]: add IMAGE of [Non-local Neural Networks](https://arxiv.org/abs/1711.07971)

File: configs/recognition/c2d/README.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+For more details on data preparation, you can refer to
+
+- [preparing_ucf101](/tools/data/ucf101/README.md)
+- [preparing_kinetics](/tools/data/kinetics/README.md)
+- [preparing_sthv1](/tools/data/sthv1/README.md)
+- [preparing_sthv2](/tools/data/sthv2/README.md)
+- [preparing_mit](/tools/data/mit/README.md)
+- [preparing_mmit](/tools/data/mmit/README.md)
+- [preparing_hvu](/tools/data/hvu/README.md)
+- [preparing_hmdb51](/tools/data/hmdb51/README.md)
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:48]: For more details on data preparation, you can refer to [preparing_kinetics](/tools/data/kinetics/README.md).

File: configs/recognition/c2d/README.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+## Train
+
+You can use the following command to train a model.
+
+```shell
+python tools/train.py ${CONFIG_FILE} [optional arguments]
+```
+
+Example: train TSN model on Kinetics-400 dataset in a deterministic option.
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:58]: use c2d 😆

File: configs/recognition/c2d/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+You can use the following command to test a model.
+
+```shell
+python tools/test.py ${CONFIG_FILE} ${CHECKPOINT_FILE} [optional arguments]
+```
+
+Example: test TSN model on Kinetics-400 dataset and dump the result to a pkl file.
+
+```shell
+python tools/test.py configs/recognition/tsn/tsn_imagenet-pretrained-r50_8xb32-1x1x3-100e_kinetics400-rgb.py \
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:78]: use c2d 😆

File: configs/recognition/c2d/README.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+|          8x8x1          | MultiStep | short-side 320 |  8   | ResNet50 (original) | ImageNet |  73.16   |  90.88   | [67.2](https://github.com/facebookresearch/SlowFast/blob/main/MODEL_ZOO.md#kinetics-400-and-600) | [87.8](https://github.com/facebookresearch/SlowFast/blob/main/MODEL_ZOO.md#kinetics-400-and-600) | 10 clips x 3 crop |            x            |   21547    | [config](/configs/recognition/c2d/c2d_nopool_imagenet-pretrained-r50_8xb32-8x8x1-100e_kinetics400-rgb.py) | [ckpt TODO](TODO) | [log TODO](TODO) |
+|          8x8x1          | MultiStep | short-side 320 |  8   | ResNet50 (with temporal pooling) | ImageNet |  71.78   |  89.90   | [71.9](https://github.com/facebookresearch/video-nonlocal-net#modifications-for-improving-speed) | [90.0](https://github.com/facebookresearch/video-nonlocal-net#modifications-for-improving-speed) | 10 clips x 3 crop |            x            |   17006    | [config](/configs/recognition/c2d/c2d_imagenet-pretrained-r50_8xb32-8x8x1-100e_kinetics400-rgb.py) | [ckpt TODO](TODO) | [log TODO](TODO) |
+
+1. The values in columns named after "reference" are the results reported in the original repo, using the same model settings.
+
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:32]: no blank line here

File: configs/recognition/c2d/README.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+2. The **gpus** indicates the number of gpu we used to get the checkpoint. It is noteworthy that the configs we provide are used for 8 gpus as default.
+   According to the [Linear Scaling Rule](https://arxiv.org/abs/1706.02677), you may set the learning rate proportional to the batch size if you use different GPUs or videos per GPU,
+   e.g., lr=0.01 for 4 GPUs x 2 video/gpu and lr=0.08 for 16 GPUs x 4 video/gpu.
+
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:36]: no blank line here

File: configs/recognition/c2d/README.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+|          8x8x1          | MultiStep | short-side 320 |  8   | ResNet50 (with temporal pooling) | ImageNet |  71.78   |  89.90   | [71.9](https://github.com/facebookresearch/video-nonlocal-net#modifications-for-improving-speed) | [90.0](https://github.com/facebookresearch/video-nonlocal-net#modifications-for-improving-speed) | 10 clips x 3 crop |            x            |   17006    | [config](/configs/recognition/c2d/c2d_imagenet-pretrained-r50_8xb32-8x8x1-100e_kinetics400-rgb.py) | [ckpt TODO](TODO) | [log TODO](TODO) |
+
+1. The values in columns named after "reference" are the results reported in the original repo, using the same model settings.
+
+2. The **gpus** indicates the number of gpu we used to get the checkpoint. It is noteworthy that the configs we provide are used for 8 gpus as default.
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:33]: replace this note with: The **gpus** indicates the number of gpus we used to get the checkpoint. If you want to use a different number of gpus or videos per gpu, the best way is to set `--auto-scale-lr` when calling `tools/train.py`, this parameter will auto-scale the learning rate according to the actual batch size and the original batch size. Refer to [swin](https://github.com/open-mmlab/mmaction2/pull/1939/files#diff-1c10da6b25716e7db3b38da277f02d45dec701d289c48681cb8b4230bf3454b9).

File: configs/recognition/c2d/c2d_nopool_imagenet-pretrained-r50_8xb32-8x8x1-100e_kinetics400-rgb.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        type=dataset_type,
+        ann_file=ann_file_val,
+        data_prefix=dict(video=data_root_val),
+        pipeline=test_pipeline,
+        test_mode=True))
+
+val_evaluator = dict(type='AccMetric')
+test_evaluator = val_evaluator
+
+default_hooks = dict(checkpoint=dict(interval=3, max_keep_ckpts=3))
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/c2d_nopool_imagenet-pretrained-r50_8xb32-8x8x1-100e_kinetics400-rgb.py:94]: add this param to enable  `auto-scale-lr` in [mmengine.](https://github.com/open-mmlab/mmengine/blob/main/mmengine/runner/runner.py#L892).
```python
# Default setting for scaling LR automatically
#   - `enable` means enable scaling LR automatically
#       or not by default.
#   - `base_batch_size` = (8 GPUs) x (32 samples per GPU).
auto_scale_lr = dict(enable=False, base_batch_size=256)
```

File: configs/recognition/c2d/c2d_imagenet-pretrained-r50_8xb32-8x8x1-100e_kinetics400-rgb.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        type=dataset_type,
+        ann_file=ann_file_val,
+        data_prefix=dict(video=data_root_val),
+        pipeline=test_pipeline,
+        test_mode=True))
+
+val_evaluator = dict(type='AccMetric')
+test_evaluator = val_evaluator
+
+default_hooks = dict(checkpoint=dict(interval=3, max_keep_ckpts=3))
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/c2d_imagenet-pretrained-r50_8xb32-8x8x1-100e_kinetics400-rgb.py:94]: add `auto-scale-lr`.

File: tests/models/backbones/test_c2d.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (Dai-Wenxun) [inline tests/models/backbones/test_c2d.py:1]: add ut for R101 if applicable.

File: configs/recognition/c2d/README.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+1. The values in columns named after "reference" are the results reported in the original repo, using the same model settings.
+2. The **gpus** indicates the number of gpus we used to get the checkpoint. If you want to use a different number of gpus or videos per gpu, the best way is to set `--auto-scale-lr` when calling `tools/train.py`, this parameter will auto-scale the learning rate according to the actual batch size and the original batch size.
+3. The validation set of Kinetics400 we used consists of 19796 videos. These videos are available at [Kinetics400-Validation](https://mycuhk-my.sharepoint.com/:u:/g/personal/1155136485_link_cuhk_edu_hk/EbXw2WX94J1Hunyt3MWNDJUBz-nHvQYhO9pvKqm6g39PMA?e=a9QldB). The corresponding [data list](https://download.openmmlab.com/mmaction/dataset/k400_val/kinetics_val_list.txt) (each line is of the format 'video_id, num_frames, label_index') and the [label map](https://download.openmmlab.com/mmaction/dataset/k400_val/kinetics_class2ind.txt) are also available.
+
+For more details on data preparation, you can refer to [preparing_kinetics](/tools/data/kinetics/README.md)
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:46]: add a period.

Comment (InvincibleWyq) [reply configs/recognition/c2d/README.md]: Done

File: mmaction/models/backbones/c2d.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+@MODELS.register_module()
+class C2D(ResNet):
+    """C2D backbone.
+
+    Comparing to ResNet-50, another temporal-pool is added after the first
+    bottleneck. Detailed structure is kept same as "video-nonlocal-net" repo.
+    Please refer to https://github.com/facebookresearch/video-nonlocal-net/blob
+    /main/configs/DBG_kinetics_resnet_4gpu_c2d_nonlocal_300k.yaml
+    Please note that there are some improvement comparing to "Non-local Neural
```
Comment (Dai-Wenxun) [inline mmaction/models/backbones/c2d.py:20]:
```suggestion
    Please note that there are some improvements compared to "Non-local Neural
```

File: mmaction/models/backbones/c2d.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    Networks" paper (https://arxiv.org/abs/1711.07971).
+    Differences are noted at https://github.com/facebookresearch/video-nonlocal
+    -net#modifications-for-improving-speed
```
Comment (Dai-Wenxun) [inline mmaction/models/backbones/c2d.py:23]: add a period

File: tests/models/backbones/test_c2d.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+import torch
+
+from mmaction.models import C2D
+from ..base import generate_backbone_demo_inputs
+
+
+def test_c2d_backbone():
+    """Test c2d_r50 backbone."""
```
Comment (Dai-Wenxun) [inline tests/models/backbones/test_c2d.py:9]:
```suggestion
    """Test c2d backbone."""
```

File: configs/recognition/c2d/README.md [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+You can use the following command to train a model.
+
+```shell
+python tools/train.py ${CONFIG_FILE} [optional arguments]
+```
+
+Example: train C2D model on Kinetics-400 dataset in a deterministic option with periodic validation.
+
+```shell
+python tools/train.py configs/recognition/c2d/c2d_r50-in1k-pre_8xb32-8x8x1-100e_kinetics400-rgb.py  \
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:48]: wrong config file

File: configs/recognition/c2d/README.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+You can use the following command to test a model.
+
+```shell
+python tools/test.py ${CONFIG_FILE} ${CHECKPOINT_FILE} [optional arguments]
+```
+
+Example: test C2D model on Kinetics-400 dataset and dump the result to a pkl file.
+
+```shell
+python tools/test.py configs/recognition/c2d/c2d_r50-in1k-pre_8xb32-8x8x1-100e_kinetics400-rgb.py \
```
Comment (Dai-Wenxun) [inline configs/recognition/c2d/README.md:65]: wrong config file

**Comment by InvincibleWyq** [comment]
@Dai-Wenxun Hi,
1. CKPTs & LOGs are now available at "/mnt/petrelfs/share_data/wangyiqin/c2d/", please help upload them and generate the "download.openmmlab.com" link.
2. "params" of each model is not set yet. How can I find the params of these models?
